M2/Macaulay2/packages/Msolve.m2

```diff
@@ -0,0 +1,527 @@
+msolvePresent := run ("type msolve >/dev/null 2>&1") === 0
+-- msolveVersion := if msolvePresent then ... -- no way to get version info!!!
+msolveVersionNeeded := "0.6.4"
+msolvePresentAndModern := msolvePresent -- and match("^[0-9.]+$",msolveVersion) and msolveVersion >= msolveVersionNeeded
+
+newPackage(
+	"Msolve",
+	Version => "1.24.05", 
+    	Date => "July 2024",
+    	Authors => {{Name => "Martin Helmer", 
+		  Email => "mhelmer@ncsu.edu", 
+		  HomePage => "http://martin-helmer.com/"}, {Name => "Mike Stillman", 
+		  Email => "mike@math.cornell.edu", 
+		  HomePage => "https://math.cornell.edu/michael-e-stillman"},{Name => "Anton Leykin", 
+		  Email => "leykin@math.gatech.edu", 
+		  HomePage => "https://antonleykin.math.gatech.edu/"}},
+    	Headline => "An interface to the msolve package (https://msolve.lip6.fr/) which computes Groebner Basis and does real root isolation",
+    	AuxiliaryFiles => true,
+	DebuggingMode => false,       
+	OptionalComponentsPresent => msolvePresent
+	);
+msolveEXEwithOpts:=(opt)->"msolve -v "|opt#"level of verbosity"|" -t "|toString(opt#"number of threads");
+export{
+    "msolveGB",
+    "msolveSaturate",
+    "msolveEliminate",
+    "msolveRUR",
+    "msolveLeadMonomials",
+    "msolveRealSolutions"
+    }
+importFrom_Core { "raw", "rawMatrixReadMsolveFile" }
+
+use'readMsolveOutputFile := true;
+readMsolveOutputFile = method()
+readMsolveOutputFile(Ring,String) := Matrix => (R,mOut) -> if use'readMsolveOutputFile 
+    then map(R, rawMatrixReadMsolveFile(raw R, mOut)) else (
+	msolveOut := get mOut;
+    	moutStrL:=separate(",",first separate("[]]",last separate("[[]",msolveOut)));
+    	--the line below should be replaced by a call to the C-function to parse the string
+    	M2Out:=for s in moutStrL list value(s);
+    	matrix {M2Out};
+    	)
+
+inputOkay=method(TypicalValue=>Boolean);
+inputOkay(Ideal):=I->(
+    R:=ring I;
+    if not instance(R,PolynomialRing) then (print "input must be in a polynomial ring over a field";return false;);
+    kk:=coefficientRing(R);
+    if instance(kk,InexactFieldFamily) then(
+        print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	return false;
+	);
+    if char(kk)>0 then (
+	if (char(kk)<2^16) or (char(kk)>2^32) then(
+	    print "input must be over the rationals or a (prime) finite field of chacteristic between 2^16 and 2^32";
+	    return false;
+	    );
+	);
+    return true;
+    );
+msolveGB=method(TypicalValue=>Matrix, Options=>{"level of verbosity"=>0, "number of threads"=>maxAllowableThreads});
```

Comment:
There's already a `Verbosity` symbol -- should that be used instead?  Also, there's a `Threads` symbol defined in `FastMinors` for the same purpose (and also used in `TerraciniLoci`).  Maybe that could be moved to `Core` and used here?

Reply:
Agree. Is there anything to replace "type of output"?

Reply:
Maybe `Outputs`?

Reply:
`flattenRing` has [`Result`](https://macaulay2.com/doc/Macaulay2/share/doc/Macaulay2/Macaulay2Doc/html/_flatten__Ring_lp..._cm__Result_eq_gt..._rp.html), which I don't like very much. There should be different functions returning different types of output instead.

Reply:
Also, maybe just use Verbose so Verbosity can eventually be removed (see https://github.com/Macaulay2/M2/issues/273)

Reply:
One more thing: maybe you should be using `allowableThreads` instead of `maxAllowableThreads`? Or at least subtract a couple to leave room for I/O?

Reply:
Often maxAllowableThreads is really still too low (i.e. is less than half of the threads on the machine)... at least this has been my experience.

Reply:
I still don't think this should be the default (especially given that msolve's own default is 1!). Using this all of this many threads can cause IO problems on systems, hence why `allowableThreads` is lower by default.

On machines with more than 16 cores, `maxAllowableThreads` is capped at 16 here:
https://github.com/Macaulay2/M2/blob/ec9e9ac60ed4a8e791448942202f077a88a87e15/M2/Macaulay2/system/supervisor.cpp#L13-L21

I think the idea is that this is a good default for most people, and those running M2 on a supercomputer should edit this code and compile M2 themselves (though I would be in favor of having an environment variable or command line argument set this number).

Reply:
I would also like to have an environment variable or command line argument to set maxAllowableThreads ... I have a workstation with 32 cores (64 threads) and don't usually compile M2 from source (and wouldn't have known which .cpp file to change to even if I was compiling from source). And I imagine there are others that also have workstations but don't compile from source. 

And ya, we can make the default allowableThreads instead.